core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java

```diff
@@ -193,6 +199,52 @@ public void testFoundAllocationAndAllocating() {
     }
 
     /**
+     * Tests that when the nodes with prior allocations of the given shard all return a decision of NO,
+     * {@link AllocationDecider#canForceAllocatePrimary(ShardRouting, RoutingNode, RoutingAllocation)}
+     * is taken to account before forcing allocation to one of the NO nodes.
+     */
+    public void testForceAllocatePrimaryOnNoDecision() {
+        testAllocator.addData(node1, ShardStateMetaData.NO_VERSION, "allocId1", randomBoolean());
+
+        // tests that we force allocate the primary when canForceAllocatePrimary returns YES for all deciders
+        AllocationDeciders deciders = new AllocationDeciders(Settings.EMPTY, new AllocationDecider[] {
+            new TestAllocateDecision(Decision.NO), new ForcePrimaryDecider(Decision.YES)
+        });
+        RoutingAllocation allocation = routingAllocationWithOnePrimaryNoReplicas(deciders, false, Version.CURRENT, "allocId1");
+        boolean changed = testAllocator.allocateUnassigned(allocation);
+        assertTrue(changed);
+        assertTrue(allocation.routingNodes().unassigned().ignored().isEmpty());
+        assertEquals(allocation.routingNodes().shardsWithState(ShardRoutingState.INITIALIZING).size(), 1);
+        assertEquals(allocation.routingNodes().shardsWithState(ShardRoutingState.INITIALIZING).get(0).currentNodeId(), node1.getId());
+
+        // tests that we do not force allocate the primary when at least one decider returns NO or THROTTLE for canForceAllocatePrimary
```

Comment:
can we have different test methods here for the different parts that are test here?